File: mobile/src/main/java/com/novoda/materialised/hackernews/asynclistview/SingleTypeAdapter.java
package com.novoda.materialised.hackernews.asynclistview;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import org.jetbrains.annotations.NotNull;

final class SingleTypeAdapter<T extends ViewData<Integer>,
        V extends View & ModelledView<T>>
        extends RecyclerView.Adapter<ModelledViewHolder<V>> {
    private final List<ViewModel<T>> viewModels;
    private final ModelledViewInflater<V> viewInflater;

    SingleTypeAdapter(ModelledViewInflater<V> viewInflater) {
        this.viewModels = new ArrayList<>();
        this.viewInflater = viewInflater;
        setHasStableIds(true);
    }

    @Override
    public ModelledViewHolder<V> onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        V view = viewInflater.inflateUsing(parent);
        return new ModelledViewHolder<>(view);
    }

    @Override
    public void onBindViewHolder(ModelledViewHolder<V> holder, int position) {
        V view = holder.obtainHeldView();
        ViewModel<T> viewModel = viewModels.get(position);
        view.updateWith(viewModel);
    }

    @Override
    public int getItemCount() {
        return viewModels.size();
    }

    @Override
    public long getItemId(int position) {
        return idFor(position);
    }

    void updateWith(ViewModel<T> fullyPopulatedViewModel) {
        for (int i = 0; i < viewModels.size(); i++) {
            if (shouldUpdate(i, fullyPopulatedViewModel)) {
                viewModels.set(i, fullyPopulatedViewModel);
                notifyItemChanged(i);
                break;
            }
        }
    }

    private boolean shouldUpdate(int position, ViewModel<T> fullyPopulatedViewModel) {
        Integer id = idFor(position);
        Integer fullyPopulatedViewModelId = fullyPopulatedViewModel.getViewData().getId();
        return id.equals(fullyPopulatedViewModelId);
    }

    private Integer idFor(int position) {
        return viewModels.get(position).getViewData().getId();
    }
}

File: mobile/src/main/java/com/novoda/materialised/hackernews/asynclistview/SingleTypeAdapter.kt
package com.novoda.materialised.hackernews.asynclistview

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import java.util.ArrayList

internal class SingleTypeAdapter<T, V>(private val viewInflater: ModelledViewInflater<V>) :
    RecyclerView.Adapter<ModelledViewHolder<V>>() where T : UiData<Int>, V : View, V : ModelledView<T> {
    private val uiStates: MutableList<UiState<T>>

    init {
        this.uiStates = ArrayList()
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ModelledViewHolder<V> =
        ModelledViewHolder(viewInflater.inflateUsing(parent))

    override fun onBindViewHolder(holder: ModelledViewHolder<V>, position: Int) {
        val view = holder.obtainHeldView()
        val uiState = uiStates[position]
        view.updateWith(uiState)
    }

    override fun getItemCount(): Int {
        return uiStates.size
    }

    override fun getItemId(position: Int): Long {
        return idFor(position)!!.toLong()
    }

    fun updateWith(uiState: UiState<T>) {
        for (i in uiStates.indices) {
            if (shouldUpdate(i, uiState)) {
                uiStates[i] = uiState
                notifyItemChanged(i)
                return
            }
        }
        uiStates.add(uiState)
        notifyItemChanged(uiStates.size - 1)
    }

    private fun shouldUpdate(position: Int, fullyPopulatedUiState: UiState<T>): Boolean {
        val id = idFor(position)
        val fullyPopulatedViewModelId = fullyPopulatedUiState.data.id
        return id == fullyPopulatedViewModelId
    }

    private fun idFor(position: Int): Int? {
        return uiStates[position].data.id
    }
}
